src_features/signMessage/cmd_signMessage.c

```diff
@@ -1,199 +1,292 @@
+                0,
+                tmpCtx.messageSigningContext.hash,
+                32);
+    }
+    return true;
+}
 
-        init_value_str(is_data_ascii(workBuffer, dataLength));
+/**
+ * Feed the UI with new data
+ */
+static void feed_display(void) {
+    int c;
 
-    } else if (p1 != P1_MORE) {
-        THROW(0x6B00);
+    while ((unprocessed_length() > 0) && (remaining_ui_buffer_length() > 0)) {
+        c = *(char *) unprocessed_data();
+        if (isspace(c))  // to replace all white-space characters as spaces
+        {
+            c = ' ';
+        }
+        if (isprint(c)) {
+            sprintf(remaining_ui_buffer(), "%c", (char) c);
+            processed_size += 1;
+        } else {
+            if (remaining_ui_buffer_length() >= 4)  // 4 being the fixed length of \x00
+            {
+                snprintf(remaining_ui_buffer(), remaining_ui_buffer_length(), "\\x%02x", c);
+                processed_size += 1;
+            } else {
+                // fill the rest of the UI buffer spaces, to consider the buffer full
+                memset(remaining_ui_buffer(), ' ', remaining_ui_buffer_length());
+            }
+        }
     }
-    if (p2 != 0) {
-        THROW(0x6B00);
+
+#ifdef TARGET_NANOS
+    if ((remaining_ui_buffer_length() == 0) && (unprocessed_length > 0)) {
+        sprintf(remaining_ui_buffer() - 3, "...");
     }
-    if ((p1 == P1_MORE) && (appState != APP_STATE_SIGNING_MESSAGE)) {
-        PRINTF("Signature not initialized\n");
-        THROW(0x6985);
+#endif
+
+    if ((remaining_ui_buffer_length() == 0) ||
+        (tmpCtx.messageSigningContext.remainingLength == 0)) {
+        if (!states.ui_started) {
+            ui_191_start();
+            states.ui_started = true;
+        } else {
+            ui_191_switch_to_message();
+        }
     }
-    if (dataLength > tmpCtx.messageSigningContext.remainingLength) {
-        THROW(0x6A80);
+
+    if ((unprocessed_length() == 0) && (tmpCtx.messageSigningContext.remainingLength > 0)) {
+        apdu_reply(0x9000);
     }
+}
 
-    cx_hash((cx_hash_t *) &global_sha3, 0, workBuffer, dataLength, NULL, 0);
-    cx_hash((cx_hash_t *) &tmpContent.sha2, 0, workBuffer, dataLength, NULL, 0);
-    tmpCtx.messageSigningContext.remainingLength -= dataLength;
+/**
+ * EIP-191 APDU handler
+ *
+ * @param[in] p1 instruction parameter 1
+ * @param[in] p2 instruction parameter 2
+ * @param[in] payload received data
+ * @param[in] length data length
+ * @return whether the handling of the APDU was successful or not
+ */
+bool handleSignPersonalMessage(uint8_t p1,
+                               uint8_t p2,
+                               const uint8_t *const payload,
+                               uint8_t length) {
+    const uint8_t *data = payload;
 
-    feed_value_str(workBuffer, dataLength, is_value_str_ascii());
+    (void) p2;
+    processed_size = 0;
+    if (p1 == P1_FIRST) {
+        if ((data = first_apdu_data(data, (uint16_t *) &length)) == NULL) {
```

Comment:
`(uint16_t *) &length` has the same undefined behaviour as first comment in this file

Reply:
I guess I'll make a temporary `uint16_t u16_length` storing the `length` and give its pointer.

Reply:
Good thing the tests catched that the length was modified then re-used further in the code.